File: server/core/src/main/java/dev/slimevr/protocol/rpc/RPCBuilder.java
package dev.slimevr.protocol.rpc;

import com.google.flatbuffers.FlatBufferBuilder;
import dev.slimevr.tracking.processor.HumanPoseManager;
import dev.slimevr.tracking.processor.config.SkeletonConfigOffsets;
import solarxr_protocol.rpc.SkeletonConfigResponse;
import solarxr_protocol.rpc.SkeletonPart;


public class RPCBuilder {

	public static int createSkeletonConfig(
		FlatBufferBuilder fbb,
		HumanPoseManager humanPoseManager
	) {
		int[] partsOffsets = new int[SkeletonConfigOffsets.values().length];

		for (int index = 0; index < SkeletonConfigOffsets.values().length; index++) {
			SkeletonConfigOffsets val = SkeletonConfigOffsets.values[index];
			int part = SkeletonPart
				.createSkeletonPart(fbb, val.id, humanPoseManager.getOffset(val));
			partsOffsets[index] = part;
		}

		int parts = SkeletonConfigResponse.createSkeletonPartsVector(fbb, partsOffsets);
		return SkeletonConfigResponse.createSkeletonConfigResponse(fbb, parts);
	}
}

File: server/core/src/main/java/dev/slimevr/protocol/rpc/RPCBuilder.kt
package dev.slimevr.protocol.rpc

import com.google.flatbuffers.FlatBufferBuilder
import dev.slimevr.tracking.processor.HumanPoseManager
import dev.slimevr.tracking.processor.config.SkeletonConfigOffsets
import solarxr_protocol.rpc.SkeletonConfigResponse
import solarxr_protocol.rpc.SkeletonPart

fun createSkeletonConfig(
	fbb: FlatBufferBuilder,
	humanPoseManager: HumanPoseManager,
): Int {
	val partsOffsets = IntArray(SkeletonConfigOffsets.entries.size)

	for (index in SkeletonConfigOffsets.entries.toTypedArray().indices) {
		val `val` = SkeletonConfigOffsets.values[index]
		val part = SkeletonPart
			.createSkeletonPart(fbb, `val`.id, humanPoseManager.getOffset(`val`))
		partsOffsets[index] = part
	}

	val parts = SkeletonConfigResponse.createSkeletonPartsVector(fbb, partsOffsets)
	val userHeight = humanPoseManager.userHeightFromConfig
	return SkeletonConfigResponse.createSkeletonConfigResponse(fbb, parts, userHeight)
}
